server/src/test/java/umm3601/user/UserControllerSpec.java

```diff
@@ -203,6 +211,38 @@ public void GetUsersByAge() throws IOException {
     }
   }
 
+  @Test
+  public void getUsersByAge() throws JsonMappingException, JsonProcessingException {
+    Context ctx = mock(Context.class);
+    // When the controller calls `ctx.queryParamMap`, return the expected map for an
+    // "?age=37" query.
+    when(ctx.queryParamMap()).thenReturn(Map.of(UserController.AGE_KEY, List.of("37")));
+    // When the controller calls `ctx.queryParamAsClass() to get the value associated with
+    // the "age" key, return an appropriate Validator. TBH, I never did figure out what the
+    // third argument to the Validator constructor was for, but `null` seems OK. I'm also not sure
+    // what the first argument is; it appears that you can set it to anything that isn't
+    // null and it's happy.
+    Validator<Integer> validator = new Validator<Integer>("age", 37, null);
+    when(ctx.queryParamAsClass(UserController.AGE_KEY, Integer.class)).thenReturn(validator);
+
+    // Call the method under test.
+    userController.getUsers(ctx);
+
+    // Verify that `getUsers` called `ctx.status(200)` at some point.
+    verify(ctx).status(200);
+
+    // Verify that `ctx.json()` is called with a `List` of `User`s.
+    // Each of those `User`s should have age 37.
+    verify(ctx).json(argThat(new ArgumentMatcher<List<User>>() {
+      public boolean matches(List<User> users) {
+        for (User user : users) {
+          assertEquals(37, user.age);
+        }
+        return true;
+      }
+    }));
```

Comment:
I think this is a test of the controller. We're saying that if the request has the age being 37, then all the results should be 37. That does indirectly test the database, but it also requires that we make the right query to the database, etc., which is what this really tests.

We could actually mock the DB as well if we wanted to take it completely out of the unit testings loop, but that would be a pretty substantial change.

Reply:
This PR has been updated to include this new test along with the other new tests. I kept this one even though it's a bit different at least in part because of the use of the argument matcher.